### ISSUE TITLE
Feature Request: Expose resetGroupIds from GroupIDAllocator

### ISSUE THREAD

**Issue by kentcdodds** [styled-components/styled-components#3660]
I'm trying to find a good way to server render without needing to use the babel plugin. I've noticed that in my example I'm working on that the first server render hydrates without issue, but subsequent server renders have a different group number for the stylesheet and that changes the hash which results in hydration errors. Here's what I'm doing to get this:

```tsx
// app/entry.server.tsx
import ReactDOMServer from "react-dom/server";
import { RemixServer } from "remix";
import type { EntryContext } from "remix";
import { ServerStyleSheet } from "styled-components";

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const sheet = new ServerStyleSheet();

  let markup = ReactDOMServer.renderToString(
    sheet.collectStyles(
      <RemixServer context={remixContext} url={request.url} />
    )
  );
  const styles = sheet.getStyleTags();

  markup = markup.replace("__STYLES__", styles);

  responseHeaders.set("Content-Type", "text/html");

  return new Response("<!DOCTYPE html>" + markup, {
    status: responseStatusCode,
    headers: responseHeaders
  });
}

// app/root.tsx
export default function App() {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width,initial-scale=1" />
        <Meta />
        <Links />
        {typeof document === "undefined" ? "__STYLES__" : null}
      </head>
      <body>
        <Outlet />
        <ScrollRestoration />
        <Scripts />
        {process.env.NODE_ENV === "development" && <LiveReload />}
      </body>
    </html>
  );
}
```

So all I would need is for the `resetGroupIds` in `GroupIDAllocator` to be exported and then I could do this:

```tsx
// app/entry.server.tsx
// ...
import { ServerStyleSheet, resetGroupIds } from "styled-components";

export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const sheet = new ServerStyleSheet();

  let markup = ReactDOMServer.renderToString(
    sheet.collectStyles(
      <RemixServer context={remixContext} url={request.url} />
    )
  );
  const styles = sheet.getStyleTags();
  resetGroupIds();

  markup = markup.replace("__STYLES__", styles);

  // ...
}
```

Would this be possible?

**Comment by kitten** [comment]
Hm, I'm going to be extremely honest here — I haven't thought about the grouping and ordering in the internal sheet code in months, so my thoughts on this may not be entirely accurate.

The groups are registered as components are created; that's by design, since we're registering the order in which they were created rather than the order of their first render call. (or first effect for that matter)

So, when these IDs become inconsistent and change, that's often an indication that you aren't dealing with the same instances of components anymore and they've been recreated, or the modules have been replaced.

The reason that's relevant is that if they were to just be reset, there's no way to actually get back to a state where they're instantiated in creation order, unless the components are actually fully recreated.

Now, again, I haven't thought about this in a while but some time ago I actually wanted to figure out some more tricks to make `componentId` for SSR obsolete, and make this overall more consistent. I simply don't have much time allocated for myself on styled-components anymore 

What's interesting here though is that we may want to first check how these IDs are changing. Specifically, if they all change at the same time (i.e. all components are recreated, is the only difference the minimum number now starting at the last maximum ID +1?). If so then we'd only have to ensure that for SSR we always adjust all of them by subtracting the minimum.

**Comment by kentcdodds** [comment]
Makes sense I think. The fact is that we don't really need to worry or care about order inconsistencies for the server render because there are no re-renders on the server. So we just want to simulate the same experience on the server as the client has for its initial render, which means: no existing groups.

I think exposing this function is all that's needed. Getting rid of the groups seems more dangerous and would require more time. I think exposing this function would be a great stepping stone and unblock folks who want to do SSR without babel.

**Comment by kitten** [comment]
Well, it depends on whether any rehydration happens on the client-side because currently we basically use server-ordering to initialize client-ordering. Admittedly, we can maybe reduce that somehow, but the problem is then still that you need to associate which components are the same on client and server side, hence the Babel plugin's component IDs. However, one idea I had was to associate them using heuristics and render ordering but then to leave the group IDs independent in both places. Either way, it's a little tricky

**Comment by kentcdodds** [comment]
Can we not assume that the first client render will be the same as the first server render? I think that's a safe assumption since otherwise you get a hydration warning.

**Comment by kitten** [comment]
So, to just write this down, I think what I was thinking about last, in terms of what could work was to change the sheet logic to this:

1. server-side: components are created and assigned a group ID (i.e. 1 then 2 then 3, as components are created)
2. server-side: components render, styles are inserted as per group ID order, but we store them with their `useId` output
3. ssr result is sent to the client-side
4. client-side: components are created and assigned a group ID (i.e. 1 then 2 then 3, as components are created)
5. client-side: logic looks at the ssr styles result and maps "old" groups in the styles to `useId` output
6: client-side: starts rendering/rehydrating with the ssr result
7: client-side: as a component renders it uses `useId` and deletes & replaces styles in the SSR styles before mounting new styles

This would mean that the client- and server-styles are completely independent from one another but we can still associate them and delete server-styles as needed at the correct time.

As long as we then use `useInsertionEffect`, in theory (!), there should never be a tick where styles are conflicting.

**Comment by kentcdodds** [comment]
Sounds like that could work for us! I'll update the issue title to describe what I'm really looking for rather than the specific implementation.

**Comment by kentcdodds** [comment]
I'm fine with whatever you decide. Right now, [the styled-components example on Remix](https://github.com/remix-run/remix/blob/dev/examples/styled-components) works. It just has a hydration warning during development after the first page is requested. It's annoying, but it works. I believe that what you're proposing could resolve this issue.

**Comment by mohammad-mustafa-24ourx** [comment]
creating remix scretch project, and getting this kind of below warning message initially really hurts the developer learning curve and interest.

> react-dom.development.js:67 Warning: Expected server HTML to contain a matching <head> in <html>.

This is why I moved to svelte from remix.

**Comment by iDVB** [comment]
@kentcdodds @kitten 
Not sure where this was left? 
Can anyone share any updates?
We've established a migration path to more to Remix, and keep our custom react ui to use styled-components, however, seeing those dev errors gives us pause. We don't want to bring back babel-plugin-styled-components hydration issues we saw in the past.

Does anyone have a sense on if these is just "dev" errors/warnings that don't physically impact anything or are they impacting production environments?

**Comment by iDVB** [comment]
Thanks @kentcdodds ! I'll try that.

**Comment by Xiphe** [comment]
I have a slightly related issue with remix+styled components where in production the styles from styled-components are missing completely => https://github.com/remix-run/examples/issues/136

_Leaving it here since people might have the same issue but land on this one_

**Comment by jmurzy** [comment]
> As a workaround, you can use [`patch-package`](https://npm.im/patch-package) to export the `resetGroupIds` in `GroupIDAllocator` and do what I suggest in my opening comment and you'll be in business. It would be a pretty low-risk thing to do I think since it's a very simple change.

Looks like `resetGroupIds` is intended for testing and dead code eliminated from the release bundle. So patching isn't an option. The only other option is to fork, export & build.

**Comment by iDVB** [comment]
@jmurzy Gah! Back to looking into the fix for this. 

@kentcdodds Can you clearify what you say is working vs not with this issue?

Our example is that the remix site loads up perfectly fine with all the styles **on first load** however, if we refresh (even with JS disabled) the second load the hashes get all misaligned and thus the styles are broken.

Is this the same issue?

**Comment by kentcdodds** [comment]
It's been a long time and I don't use styled-components. I was just trying to help others. I'm afraid I don't have any time to remind myself what this issue is about. Sorry.

**Comment by iDVB** [comment]
Nw @kentcdodds I know how it is. 
However, what may still be relevant is that Styled-Components, as is, still does not work with Remix, or any other SSR without the aid of `babel-plugin-styled-components` which itself is not supported in Remix since it doesnt seem we have access to plugins or modifying esbuild config. 
 
As a comparison, Next.js allows you to control webpack option values from a configuration file (next.config.js) so they can implement the needed plugins.

This limitation seems to be leading some remix users to [implement workarounds](https://mortenbarklund.com/blog/styled-components-remix) using a [custom module to override esbuild](https://github.com/aiji42/remix-esbuild-override).

We're VERY much wanting to move to Remix but need to slowly migrate away from styled-components, and not cold-turkey cut it off. So we need a migration path inside of Remix.

**Comment by kentcdodds** [comment]
The `patch-package` process described above worked fine for me. I'm afraid this is just not a priority for me. I'm no longer working at Remix either so spending time helping people adopt Remix isn't a priority for me either. Sorry.

**Comment by iDVB** [comment]
Makes sense @kentcdodds, no worries. Can you describe how you got patchpackage to work given what @jmurzy [says above](https://github.com/styled-components/styled-components/issues/3660#issuecomment-1416185601)... about all the "resetGroupIds" code being stripped out of the npm module as it looks like they only use it in dev?

**Comment by kentcdodds** [comment]
It's possible I remember wrong and I actually never got it working 😵‍💫 sorry I can't be more helpful. Feels like a fork may be in order. This package isn't super active, especially if you're planning on migrating away.

**Comment by roj1512** [comment]
I’ve created a CSS-in-JS library that builds CSS like Tailwind, and doesn’t require a Babel/esbuild plugin. Specifically for having no issues in Remix.

Read more here: https://github.com/roj1512/classno
Example app: https://github.com/roj1512/classno_remix

**Comment by quantizor** [comment]
Closing — this issue is stale (2+ years without resolution) and relates to an older version or has been superseded by v6.x changes. If this is still reproducible on the latest v6.x release, please open a new issue with a minimal reproduction.
